### PR TITLE
SDR Metrics classes

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
Implementation for #155 

So far just stubs.

### CODE REVIEW

**Comment by ctrl-z-9000-times** [comment]
For reference: [Standard Deviation of an Exponentially-weighted Mean](https://stats.stackexchange.com/questions/111851/standard-deviation-of-an-exponentially-weighted-mean)

File: src/nupic/ntypes/Sdr.hpp [hunk 1]
```diff
@@ -1035,8 +1088,8 @@ class SDR_ActivationFrequency {
     Real mean();
     Real std();
     Real entropy();
-    String pretty_print(); // Uses all the metrics.
-}
+    void print(std::ostream &stream = std::cout); // Uses all the metrics.
```
Comment (breznak) [inline src/nupic/ntypes/Sdr.hpp]: override << ?

**Comment by ctrl-z-9000-times** [comment]
- [ ] TODO: I think that the SDR-Metrics should be split off into their own file.  The SDR file is getting too long ( > 1000 lines )

- [ ] TODO: I'd like the SDR_metrics to have another constructor which does not accept an SDR, instead the user must call `Metric.addData( SDR )`.  This lets the users manage their own data and is a more flexible solution.

**Comment by ctrl-z-9000-times** [comment]
`SDR_Proxy` now has its own file & unit-test file.
`SDR_Metrics` now have their own file & unit-test file.

**Comment by dkeeney** [comment]
As long as you are doing this, I would like to make a suggestion.

In the interest of trying to put some isolation between the algorithms and engine (Network framework) lets have the 'types' and 'utils' folders be for algorithms and the 'ntypes' be for the network framework related stuff.  The framework can use both but the algorithms should not use things in 'ntypes'.

So if we agree on this, you could move your new files to the 'types' or 'utils' folder rather than ntypes?
I will move Watcher from utils to engine.

The Array class and its base ArrayBase is a problem. I would think it is mostly in the Network framework side but it is access by algorithms.

Your thoughts?

**Comment by ctrl-z-9000-times** [comment]
> So if we agree on this, you could move your new files to the 'types' or 'utils' folder rather than ntypes?

Okay, but lets do that in a separate PR.  

Here is my opinion on how to organize these directories:
* Move all of the supporting code for the core algorithms & encoders to utils.
  - types/Serializable.hpp -> utils
  - types/Types.hpp -> utils
  - types/Exception.hpp -> utils
  - ntypes/Sdr -> utils
  - types/Fraction.hpp -> utils   ... Maybe, its not used but I think it could be useful to keep with the utils
* Move all of data types for the engine to types.
  - ntypes/* -> types

**Comment by dkeeney** [comment]
For the orgainization... I like that.  I will document this in an issue and get @breznak's agreement.

**Comment by breznak** [comment]
Great to see the Metrics PR merged! :clap:  
Back from work abroad, I'll be catching up with the progress..bear with me :) 
And a great start to 2019! :100: :1st_place_medal:  Best of luck and lots of joy in what you do! Thanks guys!

**Pull request merged**